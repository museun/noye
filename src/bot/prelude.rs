#[doc(inline)]
pub use crate::command::{Context, Dispatcher, Event, Noye};

#[doc(inline)]
pub use template::*;

#[doc(inline)]
pub use crate::config::{self, ApiKey, Config};

#[doc(inline)]
pub use crate::irc::{Command, Target};

#[doc(inline)]
pub use crate::matches::Matches;

#[doc(inline)]
pub use crate::format::{CommaSeparated, FileSize, Iso8601, Timestamp};

use futures::prelude::*;
pub async fn concurrent_map<Iter, Item, Func, Fut, Ok, Err, Hint>(
    namespace: &str,
    hint: Hint,
    input: Iter,
    func: Func,
) -> impl Stream<Item = Ok>
where
    Iter: IntoIterator<Item = Item>,
    Func: Copy + FnOnce(Item) -> Fut,
    Fut: TryFuture<Ok = Ok, Error = Err>,
    Err: std::fmt::Display,
    Hint: Into<Option<usize>>,
{
    let (buf_tx, buf_rx) = tokio::sync::mpsc::unbounded_channel();
    stream::iter(input)
        .for_each_concurrent(hint, |item| {
            let buf_tx = buf_tx.clone();
            async move {
                if let Ok(ok) = func(item)
                    .inspect_err(|err| log::warn!("{} produced an error: {}", namespace, err))
                    .await
                {
                    let _ = buf_tx.send(ok);
                }
            }
        })
        .await;
    drop(buf_tx); // drop so the stream will end
    buf_rx
}
