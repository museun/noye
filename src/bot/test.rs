pub use crate::bot::prelude::*;
pub use crate::irc::*;

pub use futures::prelude::*;
pub use tokio::sync::mpsc;

pub fn say_template<T: Template>(context: &Context, template: T) -> String {
    let data = crate::command::resolve_template(template).unwrap();
    match context.target().unwrap() {
        Target::Channel(target) | Target::Private(target) => {
            format!("PRIVMSG {} :{}", target, data)
        }
    }
}

pub fn reply_template<T: Template>(context: &Context, template: T) -> String {
    let data = crate::command::resolve_template(template).unwrap();
    match context.target().unwrap() {
        Target::Channel(target) => format!(
            "PRIVMSG {} :{}: {}",
            target,
            context.nick().expect("nick must be attached to message"),
            data
        ),
        Target::Private(target) => format!("PRIVMSG {} :{}", target, data),
    }
}

// TODO allow for asserting on errors
pub fn check_error<F, Fut>(func: F, input: Context)
where
    F: Copy + FnOnce(Context, Noye) -> Fut,
    Fut: Future<Output = anyhow::Result<()>>,
{
    let (tx, mut rx) = mpsc::channel(32);
    let noye = Noye::new(tx);
    tokio_test::block_on(async move {
        func(input, noye).await.unwrap_err();
        assert!(rx.next().await.is_none())
    });
}

pub fn check<F, Fut>(func: F, input: Context, mut output: Vec<&str>)
where
    F: Copy + FnOnce(Context, Noye) -> Fut,
    Fut: Future<Output = anyhow::Result<()>>,
{
    let (tx, mut rx) = mpsc::channel(32);
    let noye = Noye::new(tx);

    tokio_test::block_on(async move {
        if let Err(err) = func(input.clone(), noye).await {
            panic!("failed to run '{}' on '{:#?}'", err, input.message());
        }
        let mut index = 0_usize;
        while let Some(msg) = rx.next().await {
            if output.is_empty() {
                panic!("got input: ({}) '{}' but output was empty", index, msg);
            }
            let left = output.remove(0);
            assert_eq!(
                left,
                msg,
                "expected at output pos: {}. '{}' != '{}'",
                index,
                left.escape_debug(),
                msg.escape_debug()
            );
            index += 1;
        }
    });
}
