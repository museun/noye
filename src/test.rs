#![allow(dead_code)]
pub use crate::{responses::*, *};
pub use futures::prelude::*;

use serde::{Deserialize, Serialize};
use std::sync::Arc;
use tokio::sync::Mutex;

#[derive(Default, Clone)]
pub struct YamlResponder {
    pub responses: Arc<std::sync::Mutex<Vec<Response>>>,
}

impl YamlResponder {
    pub fn len(&self) -> usize {
        self.responses.lock().unwrap().len()
    }

    pub fn is_empty(&self) -> bool {
        self.len() == 0
    }

    pub fn get_say<T>(&self) -> T
    where
        for<'de> T: Deserialize<'de>,
    {
        match self.remove() {
            Response::Say(say) => serde_yaml::from_str(&say).unwrap(),
            d => panic!("expected 'say' got {}", resp_kind(d)),
        }
    }

    pub fn get_reply<T>(&self) -> T
    where
        for<'de> T: Deserialize<'de>,
    {
        match self.remove() {
            Response::Reply(reply) => serde_yaml::from_str(&reply).unwrap(),
            d => panic!("expected 'reply' got {}", resp_kind(d)),
        }
    }

    pub fn get_raw(&self) -> String {
        match self.remove() {
            Response::Raw(raw) => raw,
            d => panic!("expected 'raw' got {}", resp_kind(d)),
        }
    }

    pub fn expect_empty(&self) {
        assert!(
            self.is_empty(),
            "expected no responses, got: {}",
            self.len()
        )
    }

    fn remove(&self) -> Response {
        let responses = &mut *self.responses.lock().unwrap();
        assert!(!responses.is_empty(), "expected a response");
        responses.remove(0)
    }
}

impl Responder for YamlResponder {
    fn say<T>(&mut self, _: Context, template: T) -> AnyhowFut<'_>
    where
        T: template::Template + Send + Sync + 'static,
        T: Serialize,
    {
        Box::pin(async move {
            self.responses
                .lock()
                .unwrap()
                .push(Response::Say(serde_yaml::to_string(&template).unwrap()));
            Ok(())
        })
    }

    fn reply<T>(&mut self, _: Context, template: T) -> AnyhowFut<'_>
    where
        T: template::Template + Send + Sync + 'static,
        T: Serialize,
    {
        Box::pin(async move {
            self.responses
                .lock()
                .unwrap()
                .push(Response::Reply(serde_yaml::to_string(&template).unwrap()));
            Ok(())
        })
    }
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub enum Response {
    Say(String),
    Reply(String),
    Raw(String),
}

fn resp_kind(response: Response) -> &'static str {
    match response {
        Response::Reply(..) => "reply",
        Response::Raw(..) => "raw",
        Response::Say(..) => "say",
    }
}

pub fn set_snapshot_path() {
    let mut settings = insta::Settings::new();
    settings.set_snapshot_path(get_snapshot_path());
    settings.bind_to_thread();
}

pub fn get_snapshot_path() -> std::path::PathBuf {
    std::path::PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("snapshots")
}

pub struct TestEnv {
    responder: YamlResponder,
    data: String,
    sender: String,
    channel: String,
    state: Arc<Mutex<State>>,
}

impl TestEnv {
    pub fn new(data: impl ToString) -> Self {
        let mut config = Config::default();
        config.irc_config.name = "test_bot".into();
        config.irc_config.channels.push("#test_channel".into());
        config.irc_config.owners.push("test_owner".into());

        let mut state = State::default();
        state.insert(CachedConfig::new(config, "noye.toml"));

        Self {
            responder: YamlResponder::default(),
            data: data.to_string(),
            sender: "test_user".into(),
            channel: "#test_channel".into(),
            state: Arc::new(Mutex::new(state)),
        }
    }

    pub fn owner(mut self) -> Self {
        self.sender = "test_owner".into();
        self
    }

    pub fn user(mut self, user: impl ToString) -> Self {
        self.sender = user.to_string();
        self
    }

    pub fn channel(mut self, channel: impl ToString) -> Self {
        self.channel = channel.to_string();
        self
    }

    pub fn insert<T: 'static + Send + Sync>(self, item: T) -> Self {
        assert!(self.state.lock().now_or_never().unwrap().insert(item));
        self
    }

    pub fn config(self, f: impl FnOnce(&mut Config)) -> Self {
        {
            let mut state = self.state.lock().now_or_never().unwrap();
            let mut config = state
                .expect_get_mut::<CachedConfig>()
                .unwrap()
                .get_mut()
                .now_or_never()
                .unwrap()
                .unwrap();
            f(&mut config);
        }
        self
    }

    pub async fn execute<H, F>(self, handler: H) -> YamlResponder
    where
        H: Handler<YamlResponder, Fut = F>,
        F: std::future::Future<Output = anyhow::Result<()>> + Send + 'static,
        F::Output: Send + 'static,
    {
        let msg = crate::Message {
            sender: self.sender,
            channel: self.channel,
            data: self.data,
        };

        let (tx, rx) = tokio::sync::mpsc::channel(32);
        match handler
            .call(
                crate::Context {
                    args: Arc::new(msg),
                    writer: crate::Writer(tx),
                    state: self.state.clone(),
                    quit: Default::default(),
                },
                self.responder.clone(),
            )
            .await
        {
            Ok(..) => {}
            Err(err) if err.is::<crate::util::DontCareSigil>() => {}
            Err(err) => {
                let err = err
                    .chain()
                    .enumerate()
                    .fold(String::new(), |mut a, (i, err)| {
                        a.extend(format!("\n[{}] --> ", i).drain(..));
                        a.extend(err.to_string().drain(..));
                        a
                    });
                panic!("{}", err)
            }
        }

        self.responder
            .responses
            .lock()
            .unwrap()
            .extend(rx.map(Response::Raw).collect::<Vec<_>>().await);

        self.responder
    }
}
