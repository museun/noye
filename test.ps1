pushd ..
cargo insta test --review --all --workspace-root noye --manifest-path .\noye\Cargo.toml
popd
