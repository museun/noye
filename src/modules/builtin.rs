use super::*;

pub(super) async fn initialize_module<R>(init: &mut ModuleInit<R>) -> anyhow::Result<()>
where
    R: Responder + Send + 'static,
{
    init.commands.add("join", join)?;
    init.commands.add("part", part)?;
    init.commands.add("uptime", uptime)?;
    init.commands.add("restart", restart)?;
    init.commands.add("respawn", respawn)?;
    init.commands.add("logs", get_logs)?;

    init.state.expect_insert(StartTime::default())
}

pub async fn join<R: Responder>(context: Context, mut responder: R) -> Result {
    context.expect_owner(&mut responder).await?;
    if let Some(chan) = context.command_args().get(0) {
        return context.writer.clone().join(chan).await;
    }
    responder.reply(context, Join::ExpectedChannel).await
}

pub async fn part<R: Responder>(mut context: Context, mut responder: R) -> Result {
    context.expect_owner(&mut responder).await?;
    let chan = &context.args.channel;
    context.writer.part(chan).await
}

pub async fn uptime<R: Responder>(context: Context, mut responder: R) -> Result {
    let uptime = context
        .state
        .lock()
        .await
        .expect_get::<StartTime>()?
        .as_readable_time();
    responder.say(context, Uptime::Uptime { uptime }).await
}

pub async fn get_logs<R: Responder>(context: Context, mut responder: R) -> Result {
    let state = context.state.lock().await;
    let temp = state.expect_get::<crate::http::server::TempStore>()?;
    let log_file = state.expect_get::<crate::LogFile>()?;
    let ExternalIp { address, port } = state.expect_get::<ExternalIp>()?.clone();

    use rand::prelude::*;

    let mut rng = rand::rngs::SmallRng::from_entropy();
    let id = temp
        .inner
        .write()
        .await
        .insert_text_file(&mut rng, log_file.0.clone())
        .await?;

    let link = format!("http://{}:{}/t/{}", address, port, id);
    let template = responses::TempStore::Link { link };
    responder.say(context.clone(), template).await
}

pub async fn restart<R: Responder>(context: Context, mut responder: R) -> Result {
    context.expect_owner(&mut responder).await?;
    let addr = context.config().await?.modules.restart.address;
    Supervisor::Restart.write(&addr).await
}

pub async fn respawn<R: Responder>(context: Context, mut responder: R) -> Result {
    context.expect_owner(&mut responder).await?;
    let delay = context
        .command_args()
        .get(0)
        .and_then(|s| s.parse().ok())
        .unwrap_or(15);
    let addr = context.config().await?.modules.restart.address;
    Supervisor::Delay(delay).write(&addr).await
}

pub struct StartTime(pub tokio::time::Instant);

impl Default for StartTime {
    fn default() -> Self {
        Self(tokio::time::Instant::now())
    }
}

impl StartTime {
    fn as_readable_time(&self) -> String {
        self.0.elapsed().as_readable_time()
    }
}

#[derive(Debug, Copy, Clone)]
enum Supervisor {
    Restart,
    Delay(u16),
}

impl Supervisor {
    async fn write(self, addr: &str) -> anyhow::Result<()> {
        use tokio::{io::*, net::TcpStream};
        let mut stream = TcpStream::connect(addr)
            .await
            .with_context(|| format!("cannot connect to restart daemon at '{}'", addr))?;

        match self {
            Self::Restart => stream.write_all(b"RESTART\0").await?,
            Self::Delay(d) => {
                stream
                    .write_all(format!("DELAY {}\0", d).as_bytes())
                    .await?
            }
        }

        stream.flush().await?;
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use crate::test::*;
    use tokio::prelude::*;

    #[tokio::test]
    async fn join_not_owner() {
        set_snapshot_path();

        let responses = TestEnv::new("!join").execute(super::join).await;
        insta::assert_yaml_snapshot!(responses.get_reply::<responses::Builtin>());
        responses.expect_empty();
    }

    #[tokio::test]
    async fn join_no_channel() {
        set_snapshot_path();

        let responses = TestEnv::new("!join").owner().execute(super::join).await;
        insta::assert_yaml_snapshot!(responses.get_reply::<responses::Join>());
        responses.expect_empty();
    }

    #[tokio::test]
    async fn join_channel() {
        set_snapshot_path();

        let responses = TestEnv::new("!join #test")
            .owner()
            .execute(super::join)
            .await;
        insta::assert_yaml_snapshot!(responses.get_raw());
        responses.expect_empty();
    }

    #[tokio::test]
    async fn part_not_owner() {
        set_snapshot_path();

        let responses = TestEnv::new("!part").execute(super::part).await;
        insta::assert_yaml_snapshot!(responses.get_reply::<responses::Builtin>());
        responses.expect_empty();
    }

    #[tokio::test]
    async fn part_channel() {
        set_snapshot_path();

        let responses = TestEnv::new("!part").owner().execute(super::part).await;
        insta::assert_yaml_snapshot!(responses.get_raw());
        responses.expect_empty();
    }

    #[tokio::test]
    async fn uptime() {
        set_snapshot_path();

        tokio::time::pause();
        let start = super::StartTime::default();
        tokio::time::advance(std::time::Duration::from_secs(10)).await;

        let responses = TestEnv::new("!uptime")
            .insert(start)
            .execute(super::uptime)
            .await;
        insta::assert_yaml_snapshot!(responses.get_say::<responses::Uptime>());
        responses.expect_empty();
    }

    #[tokio::test]
    async fn restart_not_owner() {
        set_snapshot_path();

        let mut listen = tokio::net::TcpListener::bind("localhost:0").await.unwrap();
        let addr = listen.local_addr().unwrap().to_string();

        let responses = TestEnv::new("!restart")
            .config(|config| config.modules.restart.address = addr.clone())
            .execute(super::restart)
            .await;
        insta::assert_yaml_snapshot!(responses.get_reply::<responses::Builtin>());
        responses.expect_empty();
        assert!(listen.accept().now_or_never().is_none());
    }

    #[tokio::test]
    async fn restart() {
        set_snapshot_path();

        let mut listen = tokio::net::TcpListener::bind("localhost:0").await.unwrap();
        let addr = listen.local_addr().unwrap().to_string();

        let responses = TestEnv::new("!restart")
            .config(|config| config.modules.restart.address = addr)
            .owner()
            .execute(super::restart)
            .await;
        responses.expect_empty();
        let data = accept_and_read(&mut listen).await;
        assert_eq!(data, b"RESTART\0");
    }

    #[tokio::test]
    async fn respawn_not_owner() {
        set_snapshot_path();

        let mut listen = tokio::net::TcpListener::bind("localhost:0").await.unwrap();
        let addr = listen.local_addr().unwrap().to_string();

        let responses = TestEnv::new("!respawn")
            .config(|config| config.modules.restart.address = addr.clone())
            .execute(super::respawn)
            .await;
        insta::assert_yaml_snapshot!(responses.get_reply::<responses::Builtin>());
        responses.expect_empty();
        assert!(listen.accept().now_or_never().is_none());
    }

    #[tokio::test]
    async fn respawn_default() {
        set_snapshot_path();

        let mut listen = tokio::net::TcpListener::bind("localhost:0").await.unwrap();
        let addr = listen.local_addr().unwrap().to_string();

        let responses = TestEnv::new("!respawn")
            .config(|config| config.modules.restart.address = addr.clone())
            .owner()
            .execute(super::respawn)
            .await;
        responses.expect_empty();
        let data = accept_and_read(&mut listen).await;
        assert_eq!(data, b"DELAY 15\0");
    }

    #[tokio::test]
    async fn respawn_custom() {
        set_snapshot_path();

        let mut listen = tokio::net::TcpListener::bind("localhost:0").await.unwrap();
        let addr = listen.local_addr().unwrap().to_string();

        let responses = TestEnv::new("!respawn 30")
            .config(|config| config.modules.restart.address = addr)
            .owner()
            .execute(super::respawn)
            .await;
        responses.expect_empty();
        let data = accept_and_read(&mut listen).await;
        assert_eq!(data, b"DELAY 30\0");
    }

    async fn accept_and_read(listener: &mut tokio::net::TcpListener) -> Vec<u8> {
        let (mut socket, _) = listener.accept().await.unwrap();
        let mut data = vec![];
        let _ = socket.read_to_end(&mut data).await.unwrap();
        data
    }
}
