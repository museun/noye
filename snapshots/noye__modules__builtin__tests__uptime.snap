---
source: src/modules/builtin.rs
expression: "responses.get_say::<responses::Uptime>()"
---
Uptime:
  uptime: 10 seconds
